src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let newState = { ...state };
+  const result = [];
+
+  for (const item of actions) {
+    switch (item.type) {
+      case 'addProperties':
+        newState = { ...newState, ...item.extraData };
+        result.push(newState);
+        break;
+
+      case 'removeProperties':
+        newState = { ...newState };
+
+        for (const key of item.keysToRemove) {
+          delete newState[key];
+        }
+        result.push(newState);
+        break;
+
+      case 'clear':
+        newState = {};
+        result.push(newState);
+        break;
+    }
```

Comment:
A 'default' case should be added to the switch statement for error handling, to manage or report an unsupported action type.